File: Learning_Japan_project_fe/learning-app-fe/src/services/lessonDocumentService.ts
import { http } from "@/lib/http";
import { API_ENDPOINTS } from "@/config/api";

/* ===================== TYPES ===================== */

export interface LessonDocumentResponse {
  id: string;
  title: string;
  documentUrl: string;
  documentOrder: number;
}

export interface CreateLessonDocumentRequest {
  lessonId: string;
  title: string;
  documentOrder: number;
  file: File;
}

/* ===================== SERVICE ===================== */

export const lessonDocumentService = {
  /**
   * ✅ Upload document (multipart/form-data)
   */
  create(request: CreateLessonDocumentRequest): Promise<string> {
    const formData = new FormData();
    formData.append("lessonId", request.lessonId);
    formData.append("title", request.title);
    formData.append("documentOrder", request.documentOrder.toString());
    formData.append("file", request.file);

    return http.post<string>(API_ENDPOINTS.LESSON_DOCUMENT.CREATE, formData, {
      headers: {
        "Content-Type": "multipart/form-data",
      },
    });
  },

  /**
   * ✅ Get documents by lesson
   */
  getByLesson(lessonId: string): Promise<LessonDocumentResponse[]> {
    return http.get<LessonDocumentResponse[]>(
      API_ENDPOINTS.LESSON_DOCUMENT.GET_BY_LESSON(lessonId)
    );
  },

  /**
   * ✅ Get document detail
   */
  getDetail(documentId: string): Promise<LessonDocumentResponse> {
    return http.get<LessonDocumentResponse>(
      API_ENDPOINTS.LESSON_DOCUMENT.GET_DETAIL(documentId)
    );
  },

  /**
   * ✅ Delete document
   */
  delete(documentId: string): Promise<void> {
    return http.delete<void>(API_ENDPOINTS.LESSON_DOCUMENT.DELETE(documentId));
  },
};

File: learningApp/src/main/java/com/example/learningApp/service/course/section/lesson/LessonDocumentService.java
package com.example.learningApp.service.course.section.lesson;

import com.example.learningApp.common.EntityFinder;
import com.example.learningApp.dto.response.lesson.LessonDocumentResponse;
import com.example.learningApp.entity.Lesson;
import com.example.learningApp.entity.LessonDocument;
import com.example.learningApp.repository.LessonDocumentRepository;
import com.example.learningApp.service.cloud.S3Service;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.io.IOException;
import java.util.List;

@Service
@RequiredArgsConstructor
public class LessonDocumentService {

    private final LessonDocumentRepository lessonDocumentRepository;
    private final EntityFinder finder;
    private final S3Service s3Service;

    /* ===================== CREATE ===================== */

    public String create(
            String lessonId,
            String title,
            Integer documentOrder,
            MultipartFile file
    ) throws IOException {

        Lesson lesson = finder.lessonId(lessonId);

        // ✅ Upload file lên S3
        String s3Url = s3Service.uploadLessonDocument(file, title);

        LessonDocument document = LessonDocument.builder()
                .lesson(lesson)
                .title(title)
                .documentUrl(s3Url)   // <-- gắn URL S3
                .documentOrder(documentOrder)
                .build();

        lessonDocumentRepository.save(document);

        return "Create lesson document successfully";
    }

    /* ===================== GET BY LESSON ===================== */

    public List<LessonDocumentResponse> getByLesson(String lessonId) {

        Lesson lesson = finder.lessonId(lessonId);

        return lessonDocumentRepository
                .findByLessonOrderByDocumentOrderAsc(lesson)
                .stream()
                .map(doc -> LessonDocumentResponse.builder()
                        .id(doc.getId())
                        .title(doc.getTitle())
                        .documentUrl(doc.getDocumentUrl())
                        .documentOrder(doc.getDocumentOrder())
                        .build())
                .toList();
    }

    /* ===================== GET DETAIL ===================== */

    public LessonDocumentResponse getDetail(String documentId) {

        LessonDocument document = finder.lessonDocumentId(documentId);

        return LessonDocumentResponse.builder()
                .id(document.getId())
                .title(document.getTitle())
                .documentUrl(document.getDocumentUrl())
                .documentOrder(document.getDocumentOrder())
                .build();
    }

    /* ===================== DELETE ===================== */

    public void delete(String documentId) {

        LessonDocument document = finder.lessonDocumentId(documentId);
        lessonDocumentRepository.delete(document);
    }
}

